torch/csrc/jit/passes/tensorexpr_fuser.cpp

```diff
@@ -1182,6 +1194,10 @@ void FuseTensorExprs(
   // Get rid of dead code so that we don't waste effort fusing it.
   EliminateDeadCode(graph);
 
+  if (catOptEnabled()) {
+    OptimizeConcat(graph);
+  }
+
```

Comment:
I will get to a full review of this, but the types here aren't guaranteed, the contract is that the optimization has to set up it's own guards to do this... However it's possible we dont need shapes here..

Reply:
I have added guards in the pass to ignore cases when shapes are unknown. We can probably do some of this with partial shapes. That is for a follow up though.

Reply:
the types present here aren’t “real” types, they are suggested types by profiling. They are not guaranteed to be correct. Cc @ZolotukhinM do u have the design doc of new profiling executor that explains this ?

Reply:
So, you mean we need guards in the generated code in order to use these types that are coming out of the profiling executor? 

I'm assuming there is no other place in the compilation flow where we have even partial type info. Is this a valid assumption?

Reply:
> Do u have the design doc of new profiling executor that explains this ?

No, we haven't documented this anywhere AFAIK :/

> So, you mean we need guards in the generated code in order to use these types that are coming out of the profiling executor?
> I'm assuming there is no other place in the compilation flow where we have even partial type info. Is this a valid assumption?

Types encoded in the IR at that point are types observed during profiling runs. They are not guaranteed to be the same on subsequent runs and hence optimization passes using these info are responsible for inserting appropriate runtime checks. In the fuser it is done here: https://github.com/pytorch/pytorch/blob/ba320cec6bc0e20855f8f6ce74f4248dcfafff5e/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L1170-L1173

In the end the type info is cleared from the IR: https://github.com/pytorch/pytorch/blob/ba320cec6bc0e20855f8f6ce74f4248dcfafff5e/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L581

This design kinda limits us at expressing types that are guaranteed to be the same across all runs - we discussed that in the past and @eellison had an idea of encoding profiled types as a node attribute rather than a type of a value. But we didn't do that for a reason I do not remember :)

Reply:
For the design doc, I meant the google doc u made

Reply:
I didn’t do it because we don’t have any facility in profiling executor to guarantee types, so there wasn’t much of a point in in implementing it.  It was discussed in relation to guarding autodiff graphs but I didn’t see much of a perf gain there for complexity so I didn’t do it.

Reply:
The doc is here: https://docs.google.com/document/d/1pPHKgAk5qLKXWq4bSzosmY_wltcLd0ZE_CCclwsRmjQ/edit?ts=5f1a60ea#heading=h.vkwk1deu6xpt

But I think it doesn't reflect the actual design anymore, it describes the initial design relying on type maps, which we decided not to use.

Reply:
I get the point now. It is clear that if we retain this pass in the current location in the compilation flow and use the types from profiling executor, we need guards in the code.

The alternative is to do this after the shape analysis that @eellison is working on, when we might have partial type info which might be good enough for this optimization. 

Any preference / suggestions?

Reply:
> I didn’t do it because we don’t have any facility in profiling executor to guarantee types, so there wasn’t much of a point in in implementing it. It was discussed in relation to guarding autodiff graphs but I didn’t see much of a perf gain there for complexity so I didn’t do it.

I think it might be still worth reconsidering: the TorchScript IR could be used beyond the context of PE and it might be convenient to be able to express "guaranteed by something" vs "expected from the past runs" type info. A hypothetical scenario: we specialize the graph for given input sizes (and these sizes are checked somewhere outside the JIT runtime), but then we run it with profiling executor to refine some types of intermediate values. In this scenario we could've encoded types guaranteed externally and inferred from that as types of IR values and types observed during profiling phase as attributes.

Reply:
In relation to @ZolotukhinM's point, where would the types inferred by shape analysis stored? How will that be differentiated from the type info added by the profiling executor? @eellison

Reply:
At this point I am mostly envisioning shape analysis for use in Freezing, where types are guaranteed, and as a way to prove & generate at runtime symbolic shapes in fusion/tensorexpr. 

I think it's important to invest in our IR and infrastructure that can be reused across different workflows with stuff like shape and dtype/device analysis. Once we have better facilities around propagating and proving information than we can reconsider what the execution model will look. 

I am focusing on optimizations in freezing because the ergonomics of writing passes is much better. I would like to prove out performance wins there first. In the short-midterm, I would recommend people writing optimization passes for inference to focus on that as well. That is where shapes and other information will first be added and guaranteed. 

The current executor was proposed with the idea that graph cleanup is not important, and that fusion is the only optimization that matters. While that may or may not be true, I think it's important to focus on particular optimizations we would like to enable before putting more work into the executor. And if the original thesis becomes less true, we might consider switching to a new execution model instead of incrementally adding features to the current one.

Reply:
> I am focusing on optimizations in freezing because the ergonomics of writing passes is much better. I would like to prove out performance wins there first. In the short-midterm, I would recommend people writing optimization passes for inference to focus on that as well. That is where shapes and other information will first be added and guaranteed.

Sure, that sounds like a good plan. I'm going to wait for your shape analysis pass to land and target this as an optimization after freezing.

Reply:
Removed calling this optimization from `tensorexpr_fuser`.